File: crates/adapters/interactive_brokers/src/python/historical.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! Python bindings for the Interactive Brokers historical data client.

use std::sync::Arc;

use chrono::{DateTime, Utc};
use ibapi::contracts::Contract;
use nautilus_common::live::get_runtime;
use nautilus_core::python::{to_pyruntime_err, to_pyvalue_err};
use nautilus_model::{
    data::{Bar, Data},
    identifiers::InstrumentId,
    instruments::any::InstrumentAny,
    python::{data::data_to_pycapsule, instruments::instrument_any_to_pyobject},
};
use pyo3::{prelude::*, types::PyList};

use crate::{
    historical::HistoricalInteractiveBrokersClient, python::conversion::py_list_to_contracts,
};

#[pymethods]
impl HistoricalInteractiveBrokersClient {
    #[new]
    #[allow(clippy::needless_pass_by_value)]
    fn py_new(
        instrument_provider: crate::providers::instruments::InteractiveBrokersInstrumentProvider,
        config: crate::config::InteractiveBrokersDataClientConfig,
    ) -> PyResult<Self> {
        let shared_client = get_runtime()
            .block_on(crate::common::shared_client::get_or_connect(
                &config.host,
                config.port,
                config.client_id,
                config.connection_timeout,
            ))
            .map_err(to_pyruntime_err)?;

        Ok(Self::new(
            Arc::clone(shared_client.as_arc()),
            Arc::new(instrument_provider),
        ))
    }

    fn __repr__(&self) -> String {
        format!("{self:?}")
    }

    /// Request historical bars.
    ///
    /// # Arguments
    ///
    /// * `bar_specifications` - List of bar specifications (e.g., ["1-HOUR-LAST"])
    /// * `end_date_time` - End date for bars
    /// * `start_date_time` - Optional start date
    /// * `duration` - Optional duration string (e.g., "1 D")
    /// * `contracts` - Optional list of IB contracts (dicts with symbol, sec_type, exchange, currency, etc.)
    /// * `instrument_ids` - Optional list of instrument IDs
    /// * `use_rth` - Use regular trading hours only
    /// * `timeout` - Request timeout in seconds
    #[pyo3(signature = (bar_specifications, end_date_time, start_date_time=None, duration=None, contracts=None, instrument_ids=None, use_rth=true, timeout=60))]
    #[pyo3(name = "request_bars")]
    #[allow(clippy::too_many_arguments)]
    #[allow(clippy::needless_pass_by_value)]
    fn py_request_bars<'py>(
        &self,
        py: Python<'py>,
        bar_specifications: Vec<String>,
        end_date_time: DateTime<Utc>,
        start_date_time: Option<DateTime<Utc>>,
        duration: Option<String>,
        contracts: Option<Py<PyList>>,
        instrument_ids: Option<Vec<InstrumentId>>,
        use_rth: bool,
        timeout: u64,
    ) -> PyResult<Bound<'py, PyAny>> {
        let client = self.clone();
        let bar_specs = bar_specifications;
        let duration_str = duration;

        // Convert Python contracts list to Rust Contracts
        let contracts_vec: Option<Vec<Contract>> = if let Some(py_contracts) = contracts.as_ref() {
            let py_contracts_bound = py_contracts.bind(py);
            match py_list_to_contracts(py_contracts_bound) {
                Ok(contracts) => Some(contracts),
                Err(e) => {
                    return Err(to_pyvalue_err(format!("Failed to convert contracts: {e}")));
                }
            }
        } else {
            None
        };

        pyo3_async_runtimes::tokio::future_into_py(py, async move {
            // Convert Vec<String> to Vec<&str> for the request
            let bar_specs_refs: Vec<&str> = bar_specs.iter().map(|s| s.as_str()).collect();
            let bars: Vec<Bar> = client
                .request_bars(
                    bar_specs_refs,
                    end_date_time,
                    start_date_time,
                    duration_str.as_deref(),
                    contracts_vec,
                    instrument_ids,
                    use_rth,
                    timeout,
                )
                .await
                .map_err(to_pyruntime_err)?;
            // Convert bars to Python objects
            Ok(bars)
        })
    }

    /// Request historical ticks (quotes or trades).
    ///
    /// # Arguments
    ///
    /// * `tick_type` - Type of ticks: "TRADES" or "BID_ASK"
    /// * `start_date_time` - Start date for ticks
    /// * `end_date_time` - End date for ticks
    /// * `contracts` - Optional list of IB contracts (dicts with symbol, sec_type, exchange, currency, etc.)
    /// * `instrument_ids` - Optional list of instrument IDs
    /// * `use_rth` - Use regular trading hours only
    /// * `timeout` - Request timeout in seconds
    #[pyo3(signature = (tick_type, start_date_time, end_date_time, contracts=None, instrument_ids=None, use_rth=true, timeout=60))]
    #[pyo3(name = "request_ticks")]
    #[allow(clippy::too_many_arguments)]
    #[allow(clippy::needless_pass_by_value)]
    fn py_request_ticks<'py>(
        &self,
        py: Python<'py>,
        tick_type: String,
        start_date_time: DateTime<Utc>,
        end_date_time: DateTime<Utc>,
        contracts: Option<Py<PyList>>,
        instrument_ids: Option<Vec<InstrumentId>>,
        use_rth: bool,
        timeout: u64,
    ) -> PyResult<Bound<'py, PyAny>> {
        let client = self.clone();

        // Convert Python contracts list to Rust Contracts
        let contracts_vec: Option<Vec<Contract>> = if let Some(py_contracts) = contracts.as_ref() {
            let py_contracts_bound = py_contracts.bind(py);
            match py_list_to_contracts(py_contracts_bound) {
                Ok(contracts) => Some(contracts),
                Err(e) => {
                    return Err(to_pyvalue_err(format!("Failed to convert contracts: {e}")));
                }
            }
        } else {
            None
        };

        pyo3_async_runtimes::tokio::future_into_py(py, async move {
            let data_vec: Vec<Data> = client
                .request_ticks(
                    &tick_type,
                    start_date_time,
                    end_date_time,
                    contracts_vec,
                    instrument_ids,
                    use_rth,
                    timeout,
                )
                .await
                .map_err(to_pyruntime_err)?;
            // Convert Data enum to Python objects using pycapsules
            Python::attach(|py| -> PyResult<Py<PyList>> {
                let py_list = PyList::empty(py);
                for data in data_vec {
                    let py_capsule = data_to_pycapsule(py, data);
                    py_list.append(py_capsule)?;
                }
                Ok(py_list.into())
            })
        })
    }

    /// Request instruments.
    ///
    /// # Arguments
    ///
    /// * `instrument_ids` - Optional list of instrument IDs to load
    /// * `contracts` - Optional list of IB contracts (dicts with symbol, sec_type, exchange, currency, etc.)
    #[pyo3(signature = (instrument_ids=None, contracts=None))]
    #[pyo3(name = "request_instruments")]
    #[allow(clippy::needless_pass_by_value)]
    fn py_request_instruments<'py>(
        &self,
        py: Python<'py>,
        instrument_ids: Option<Vec<InstrumentId>>,
        contracts: Option<Py<PyList>>,
    ) -> PyResult<Bound<'py, PyAny>> {
        let client = self.clone();

        // Convert Python contracts list to Rust Contracts
        let contracts_vec: Option<Vec<Contract>> = if let Some(py_contracts) = contracts.as_ref() {
            let py_contracts_bound = py_contracts.bind(py);
            match py_list_to_contracts(py_contracts_bound) {
                Ok(contracts) => Some(contracts),
                Err(e) => {
                    return Err(to_pyvalue_err(format!("Failed to convert contracts: {e}")));
                }
            }
        } else {
            None
        };

        pyo3_async_runtimes::tokio::future_into_py(py, async move {
            let instruments: Vec<InstrumentAny> = client
                .request_instruments(instrument_ids, contracts_vec)
                .await
                .map_err(to_pyruntime_err)?;
            // Convert instruments to Python objects
            Python::attach(|py| -> PyResult<Py<PyList>> {
                let py_list = PyList::empty(py);

                for instrument in instruments {
                    let py_obj =
                        instrument_any_to_pyobject(py, instrument).map_err(to_pyruntime_err)?;
                    py_list.append(py_obj)?;
                }
                Ok(py_list.into())
            })
        })
    }
}

File: nautilus_trader/adapters/interactive_brokers_pyo3/historical.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------
"""
Provides a PyO3-based historical data client for Interactive Brokers.

This adapter uses PyO3 bindings to call the Rust implementation of the Interactive
Brokers adapter, providing the same API as the Python adapter but with Rust performance.

"""

from __future__ import annotations

from datetime import UTC
from datetime import datetime

from nautilus_trader.adapters.interactive_brokers_pyo3._contracts import IBContractSpec
from nautilus_trader.adapters.interactive_brokers_pyo3._contracts import ib_contract_specs_to_dicts
from nautilus_trader.cache.transformers import transform_instrument_from_pyo3
from nautilus_trader.core.data import Data
from nautilus_trader.core.nautilus_pyo3 import InstrumentId as PyO3InstrumentId
from nautilus_trader.model.data import Bar
from nautilus_trader.model.data import capsule_to_data
from nautilus_trader.model.instruments import Instrument


try:
    from nautilus_trader.core.nautilus_pyo3.interactive_brokers import (
        HistoricalInteractiveBrokersClient as RustHistoricalInteractiveBrokersClient,
    )
except ImportError:
    RustHistoricalInteractiveBrokersClient = None


class HistoricalInteractiveBrokersClient:
    """
    Provides a PyO3-based historical data client for Interactive Brokers.

    This class wraps the Rust implementation via PyO3 bindings, providing
    the same API as the Python adapter but using the Rust implementation.

    Parameters
    ----------
    instrument_provider : InteractiveBrokersInstrumentProvider
        The instrument provider.
    config : InteractiveBrokersInstrumentProviderConfig
        Configuration for the client.

    Raises
    ------
    ImportError
        If the PyO3 bindings are not available.

    """

    def __init__(
        self,
        instrument_provider,  # InteractiveBrokersInstrumentProvider
        config,  # InteractiveBrokersInstrumentProviderConfig
    ) -> None:
        if RustHistoricalInteractiveBrokersClient is None:
            raise ImportError(
                "PyO3 bindings for Interactive Brokers are not available. "
                "Please ensure the extension module is built with the 'extension-module' feature.",
            )

        # Initialize the Rust client via PyO3
        self._rust_client = RustHistoricalInteractiveBrokersClient(
            instrument_provider._rust_provider,
            config,
        )

    @staticmethod
    def _normalize_instrument_ids(
        instrument_ids: list[str] | None,
    ) -> list[PyO3InstrumentId] | None:
        if not instrument_ids:
            return None

        return [PyO3InstrumentId.from_str(inst_id) for inst_id in instrument_ids]

    @staticmethod
    def _normalize_instruments(instruments: list[Instrument]) -> list[Instrument]:
        return [transform_instrument_from_pyo3(item) for item in instruments]

    @staticmethod
    def _normalize_bars(bars: list[Bar]) -> list[Bar]:
        if bars and not isinstance(bars[0], Bar):
            return Bar.from_pyo3_list(bars)

        return bars

    @staticmethod
    def _normalize_data(data: list[Data]) -> list[Data]:
        if data and type(data[0]).__name__ == "PyCapsule":
            return [capsule_to_data(item) for item in data]

        return data

    @staticmethod
    def _normalize_request_datetime(value: datetime, field_name: str) -> datetime:
        if value.tzinfo is None:
            return value.replace(tzinfo=UTC)

        offset = value.utcoffset()
        if offset is None or offset.total_seconds() != 0:
            raise ValueError(
                f"{field_name} must be UTC. Configure TWS / IB Gateway to use UTC timestamps.",
            )

        return value.astimezone(UTC)

    async def request_instruments(
        self,
        instrument_ids: list[str] | None = None,
        contracts: list[IBContractSpec] | None = None,
    ) -> list[Instrument]:
        """
        Request instruments from Interactive Brokers.

        Parameters
        ----------
        instrument_ids : list[str], optional
            List of instrument ID strings (e.g., "AAPL.NASDAQ").
        contracts : list[IBContract], optional
            List of IB contracts.

        Returns
        -------
        list[Instrument]
            List of loaded instruments.

        """
        contracts_dicts = ib_contract_specs_to_dicts(contracts)

        # Convert instrument ID strings to InstrumentId objects
        instrument_id_objs = self._normalize_instrument_ids(instrument_ids)

        # Call Rust client method
        result = await self._rust_client.request_instruments(
            instrument_ids=instrument_id_objs,
            contracts=contracts_dicts,
        )

        return self._normalize_instruments(result)

    async def request_bars(
        self,
        bar_specifications: list[str],
        end_date_time: datetime,
        start_date_time: datetime | None = None,
        duration: str | None = None,
        contracts: list[IBContractSpec] | None = None,
        instrument_ids: list[str] | None = None,
        use_rth: bool = True,
        timeout: int = 120,
    ) -> list[Bar]:
        """
        Request historical bars from Interactive Brokers.

        Parameters
        ----------
        bar_specifications : list[str]
            List of bar specifications (e.g., ["1-HOUR-LAST", "5-MINUTE-MID"]).
        end_date_time : datetime
            End date/time for the bars.
        start_date_time : datetime, optional
            Start date/time. Either this or duration must be provided.
        duration : str, optional
            Duration string (e.g., "1 D", "1 W"). Either this or start_date_time must be provided.
        contracts : list[IBContract], optional
            List of IB contracts.
        instrument_ids : list[str], optional
            List of instrument ID strings.
        use_rth : bool, default True
            Use regular trading hours only.
        timeout : int, default 120
            Request timeout in seconds.

        Returns
        -------
        list[Bar]
            List of historical bars.

        """
        contracts_dicts = ib_contract_specs_to_dicts(contracts)

        # Convert instrument ID strings to InstrumentId objects
        instrument_id_objs = self._normalize_instrument_ids(instrument_ids)

        end_date_time = self._normalize_request_datetime(end_date_time, "end_date_time")

        start_dt_utc = None

        if start_date_time:
            start_dt_utc = self._normalize_request_datetime(
                start_date_time,
                "start_date_time",
            )

        # Call Rust client method
        result = await self._rust_client.request_bars(
            bar_specifications=bar_specifications,
            end_date_time=end_date_time,
            start_date_time=start_dt_utc,
            duration=duration,
            contracts=contracts_dicts,
            instrument_ids=instrument_id_objs,
            use_rth=use_rth,
            timeout=timeout,
        )

        return self._normalize_bars(result)

    async def request_ticks(
        self,
        tick_type: str,
        start_date_time: datetime,
        end_date_time: datetime,
        contracts: list[IBContractSpec] | None = None,
        instrument_ids: list[str] | None = None,
        use_rth: bool = True,
        timeout: int = 120,
    ) -> list[Data]:
        """
        Request historical ticks from Interactive Brokers.

        Parameters
        ----------
        tick_type : str
            Type of ticks: "TRADES" or "BID_ASK".
        start_date_time : datetime
            Start date/time for the ticks.
        end_date_time : datetime
            End date/time for the ticks.
        contracts : list[IBContract], optional
            List of IB contracts.
        instrument_ids : list[str], optional
            List of instrument ID strings.
        use_rth : bool, default True
            Use regular trading hours only.
        timeout : int, default 120
            Request timeout in seconds.

        Returns
        -------
        list[Data]
            List of historical ticks (QuoteTick or TradeTick).

        """
        contracts_dicts = ib_contract_specs_to_dicts(contracts)

        # Convert instrument ID strings to InstrumentId objects
        instrument_id_objs = self._normalize_instrument_ids(instrument_ids)

        start_date_time = self._normalize_request_datetime(start_date_time, "start_date_time")
        end_date_time = self._normalize_request_datetime(end_date_time, "end_date_time")

        # Call Rust client method
        result = await self._rust_client.request_ticks(
            tick_type=tick_type,
            start_date_time=start_date_time,
            end_date_time=end_date_time,
            contracts=contracts_dicts,
            instrument_ids=instrument_id_objs,
            use_rth=use_rth,
            timeout=timeout,
        )

        return self._normalize_data(result)

    async def connect(self) -> None:
        """
        Maintain compatibility with the legacy historical client API.
        """
        return None


HistoricInteractiveBrokersClient = HistoricalInteractiveBrokersClient
